core/node/node_framework/src/resource/lazy_resource.rs

```diff
@@ -83,10 +83,104 @@ impl<T: Resource + Clone> LazyResource<T> {
+
+    impl Resource for TestResource {
+        fn resource_id() -> ResourceId {
+            ResourceId::new("test_resource")
+        }
+    }
+
+    fn init_lazy_resource() -> LazyResource<TestResource> {
+        let (stop_sender, _stop_receiver) = watch::channel(false);
+        LazyResource::<TestResource>::new(StopReceiver(stop_sender.subscribe()))
+    }
+
+    fn init_test_resource() -> TestResource {
+        TestResource(Arc::new(1))
+    }
+
+    fn init_resources() -> (LazyResource<TestResource>, TestResource) {
+        (init_lazy_resource(), init_test_resource())
+    }
+
+    #[tokio::test]
+    async fn test_already_provided_resource_case() {
+        let (lazy_resource, test_resource) = init_resources();
+
+        let (mut lr_for_task1, tr_for_task1) = (lazy_resource.clone(), test_resource.clone());
+
+        let task1 = tokio::spawn(async move {
+            lr_for_task1.provide(tr_for_task1).await.unwrap();
+        });
+
+        let (mut lr_for_task2, tr_for_task2) = (lazy_resource.clone(), test_resource.clone());
+
+        let task2 = tokio::spawn(async move {
+            tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+            assert_eq!(
+                lr_for_task2.provide(tr_for_task2).await,
+                Err(LazyResourceError::ResourceAlreadyProvided)
+            );
+        });
+
+        futures::future::join_all([task1, task2]).await;
+    }
+
+    #[tokio::test]
+    async fn test_successful_resolve_case() {
+        let (lazy_resource, test_resource) = init_resources();
+
+        let (mut lr_for_task1, tr_for_task1) = (lazy_resource.clone(), test_resource.clone());
+
+        let task1 = tokio::spawn(async move {
+            lr_for_task1.provide(tr_for_task1).await.unwrap();
+        });
+
+        let (lr_for_task2, tr_for_task2) = (lazy_resource.clone(), test_resource.clone());
+
+        let task2 = tokio::spawn(async move {
+            tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+            assert_eq!(lr_for_task2.resolve().await, Ok(tr_for_task2));
+        });
+
+        futures::future::join_all([task1, task2]).await;
+    }
+
+    #[tokio::test]
+    async fn test_node_shutdown_case() {
+        let (stop_sender, _stop_receiver) = watch::channel(false);
+        let lazy_resource =
+            LazyResource::<TestResource>::new(StopReceiver(stop_sender.subscribe()));
+        let test_resource = init_test_resource();
+
+        let task1 = tokio::spawn(async move {
+            tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+            stop_sender.send(true).unwrap();
+        });
+
+        let (lr_for_task2, _) = (lazy_resource.clone(), test_resource.clone());
+
+        let task2 = tokio::spawn(async move {
+            assert_eq!(
+                lr_for_task2.resolve().await,
+                Err(LazyResourceError::NodeShutdown)
+            );
+        });
+
+        futures::future::join_all([task1, task2]).await;
```

Comment:
I think here we can use just one tokio task:
1. You spawn a task where we call `resolve`, and then
2. You send the `true` through the `stop_sender`, and then
3. We `.await` the join handle where we called `resolve()`.

Additionally, we may change the semantics a bit: instead of asserting that we've returned an error _inside_ of the tokio task, we just return what we got there, and in the main test body we make sure that return value is `Ok(Err(LazyResourceError::NodeShutdown))` (`Ok` means that the task was completed successfully and didn't panic; and `Err(..)` is what we return from the task.